File: accepted/114-flatten-binary-tree-to-linked-list.py
# https://leetcode.com/problems/flatten-binary-tree-to-linked-list/
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param {TreeNode} root
    # @return {void} Do not return anything, modify root in-place instead.
    def real_flatten(self, node):
        if None == node or (None == node.left and None == node.right):
            return node
        
        cur = node
        right_head = node.right
        left = self.real_flatten(node.left)
        right = self.real_flatten(node.right)
        
        if None != node.left:
            node.right = node.left
            node.left = None
            cur = left
        if None != right_head:
            cur.right = right_head
            cur = right
        
        return cur
        
    def flatten(self, root):
        self.real_flatten(root)

File: accepted/114-flatten-binary-tree-to-linked-list.c
/* https://leetcode.com/problems/flatten-binary-tree-to-linked-list/ */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
struct TreeNode *real_flatten(struct TreeNode *pNode)
{
    if (NULL == pNode || (NULL == pNode->left && NULL == pNode->right))
        return pNode;

    struct TreeNode *pCur = pNode;
    struct TreeNode *pRightHead = pNode->right;
    struct TreeNode *pLeft = real_flatten(pNode->left);
    struct TreeNode *pRight = real_flatten(pNode->right);

    if (NULL != pNode->left)
    {
        pNode->right = pNode->left;
        pNode->left = NULL;
        pCur = pLeft;
    }

    if (NULL != pRightHead)
    {
        pCur->right = pRightHead;
        pCur = pRight;
    }

    return pCur;
}

void flatten(struct TreeNode* root)
{
    real_flatten(root);
}
